flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/FlinkSourceFilter.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink;
+
+import org.apache.flink.api.common.functions.FilterFunction;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.types.logical.RowType;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.expressions.Evaluator;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.types.Types;
+
+public class FlinkSourceFilter implements FilterFunction<RowData> {
+
+  private final Schema schema;
+  private final Expression expr;
+  private final boolean caseSensitive;
+
+  private volatile RowDataWrapper wrapper;
+  private volatile Evaluator evaluator;
+
+  public FlinkSourceFilter(Schema schema, Expression expr, boolean caseSensitive) {
+    this.schema = schema;
+    this.expr = expr;
+    this.caseSensitive = caseSensitive;
+  }
+
+  @Override
+  public boolean filter(RowData value) throws Exception {
+    if (this.wrapper == null) {
+      RowType rowType = FlinkSchemaUtil.convert(schema);
+      Types.StructType struct = schema.asStruct();
+      this.wrapper = new RowDataWrapper(rowType, struct);
```

Comment:
The wrapper is not serializable, therefore I decided to construct it when we need it.

Reply:
Is it possible to make the wrapper serializable?

Reply:
It is but it requires some more changes. I've pushed a commit that includes those changes.